src/appleseed/renderer/modeling/camera/fisheyelenscamera.cpp

```diff
@@ -0,0 +1,511 @@
+            m_focal_length = extract_focal_length(m_film_dimensions[0]);
+
+            // Extract the abscissa of the near plane from the camera parameters.
+            m_near_z = extract_near_z();
+
+            // Precompute reciprocals of film dimensions.
+            m_rcp_film_width = 1.0 / m_film_dimensions[0];
+            m_rcp_film_height = 1.0 / m_film_dimensions[1];
+
+            // Precompute pixel area.
+            const size_t pixel_count = project.get_frame()->image().properties().m_pixel_count;
+            m_pixel_area = m_film_dimensions[0] * m_film_dimensions[1] / pixel_count;
+
+            const string projection_type = m_params.get_required<string>("projection_type", "equisolid_angle");
+
+            if (projection_type == "equisolid_angle")
+                m_projection_type = EquisolidAngle;
+
+            else if (projection_type == "equidistant")
+                m_projection_type = Equidistant;
+                
+            else if (projection_type == "stereographic")
+                m_projection_type = Stereographic;
+
+            else if (projection_type == "thoby")
+                m_projection_type = Thoby;
+
+            else
+            {
+                RENDERER_LOG_ERROR(
+                    "invalid value \"%s\" for parameter \"projection_type\", "
+                    "using default value \"equisolid_angle\".",
+                    projection_type.c_str());
+                m_projection_type = EquisolidAngle;
+            }
+
+            return true;
+        }
+
+        //
+        // Transform ray direction in selected projection.
+        // Reference : 
+        //   https://wiki.panotools.org/Fisheye_Projection
+        //   https://de.wikipedia.org/wiki/Fischaugenobjektiv
+        //   http://michel.thoby.free.fr/Fisheye_history_short/Projections/Fisheye_projection-models.html
+        //
+
+        void transform_ray_dir(Vector3d& dir) const
+        {
+            // Transform the direction vector to be in the xz plane.
+            const Transformd rot = Transformd(Matrix4d::make_rotation_z(-atan2 (dir[1], dir[0])));
+            dir = rot.vector_to_parent(dir);
+
+            // Tangent of angle between vector(on xz plane) and z+ axis.
+            const double tan_theta1 = dir[0]/dir[2];
+            const double theta1 = atan(tan_theta1);
+            double theta2 = 0.0;
+
+            switch (m_projection_type) 
+            {
+                case EquisolidAngle:
+                    theta2 = 2 * asin(tan_theta1 / 2.0);
+                    break;
+                
+                case Equidistant:
+                    theta2 = tan_theta1;
+                    break;
+
+                case Stereographic:
+                    theta2 = 2 * atan(tan_theta1 / 2.0);
+                    break;
+
+                case Thoby:
+                    theta2 = asin(tan_theta1 / 1.47) / 0.713;
+                    break;
+
+                default:
+                    RENDERER_LOG_ERROR(
```

Comment:
If the only way for this to happen is a programmer error, then use an `assert()` instead. Real error messages are reserved for errors that result from user error.

Reply:
Can I handle it like this?

    default:
      assert(false);
      break; // maybe not necessary

Reply:
Yes: if you're sure that it's impossible to reach this code without an earlier programming error, then an assertion is sufficient. I usually leave the `break` as well.